src/lib/quoter.js

```diff
@@ -39,77 +125,44 @@ class Quoter {
     if (request.destinationAmount === '0') {
       throw new InvalidAmountSpecifiedError('destinationAmount must be positive')
     }
-    return yield this._quoteByAmount(request)
+    // Use the shifted curve because the `amountReverse()` is rounded down,
+    // and we don't want to lose money.
+    const liquidityQuote = yield this._quoteLiquidity(Object.assign({_shiftCurve: true}, request))
+    if (!liquidityQuote) return null
+    return Object.assign({
+      sourceAmount: liquidityQuote.liquidityCurve.amountReverse(request.destinationAmount).toString()
+    }, liquidityQuote)
   }
 
-  * _quoteByAmount (request) {
-    const destinationHoldDuration = request.destinationHoldDuration
-    const hop = this._findBestHopForAmount(request)
-    if (!hop) return null
-
-    // note the confusion between `next`, `destination`, and `final`;
-    // `finalAmount` will be renamed to `destinationAmount` later in
-    // the hopToQuote function.
-    if (hop.sourceAmount && hop.finalAmount) {
-      return hopToQuote(hop, destinationHoldDuration)
-    }
-
-    // Otherwise, ask a connector closer to the destination.
-
-    let headHop
-    const intermediateConnector = hop.destinationCreditAccount
-    // Quote by source amount
-    if (request.sourceAmount) {
-      headHop = this.tables.findBestHopForSourceAmount(
-        hop.sourceLedger, intermediateConnector, request.sourceAmount)
-    }
-    const tailQuote = yield ILQP.quoteByConnector({
-      plugin: this.ledgers.getPlugin(hop.destinationLedger),
-      connector: intermediateConnector,
-      quoteQuery: {
-        sourceAmount: request.destinationAmount === undefined ? headHop.destinationAmount : null,
-        destinationAmount: request.sourceAmount === undefined ? hop.finalAmount : null,
-        destinationAccount: request.destinationAccount,
-        destinationHoldDuration
-      }
+  /**
+   * @param {IlpAddress} sourceLedger
+   * @param {IlpAddress} destination
+   * @param {Amount} sourceAmount
+   * @returns {Object}
+   */
+  * findBestPathForSourceAmount (sourceLedger, destination, sourceAmount) {
```

Comment:
It's maybe a bit confusing that path finding ('findBestPath...') calls liquidity quoting, which in turn ends up calling the  path finding code in the ilp-routing module. But I guess we just have that level of complexity now; we can maybe try to remove layers later in a refactor.